### PR TITLE
METRON-636: Capture memory and cpu details as a part of platform-info…

### PR DESCRIPTION
**Pull request by anandsubbu**
… script

Added top commands for Linux and Mac to be captured as a part of the platform-info.sh script.

### CODE REVIEW

**Comment by anandsubbu** [comment]
Hello, checking to see if this would be a worthwhile addition to the platform-info.sh script. Please let me know.

**Comment by anandsubbu** [comment]
Hi @nickwallen , thank you for the review. 

At the outset, getting the system CPU and Memory info would help troubleshooting, especially for single node deployments which are attempted on environments with scarce resources (This [thread](https://community.hortonworks.com/questions/51666/apache-metron-single-node-installation-problems.html) for example).

My intention is in fact to capture the 'static' system information. I agree with the challenges associated with capturing the runtime information--which this script is not intended to do.

My initial approach was to dump the contents of `/proc/cpuinfo` and `/proc/meminfo`. However, for lack of the same equivalent on Mac, I resorted to getting the first few lines of `top` (again with different switches for Linux and Mac :/). 

I can still dump the output of /proc/cpuinfo and /proc/meminfo on Linux and use a different alternative for Mac. I looked as much but could not find one for Mac. Please let me know if you are aware of one.

**Comment by nickwallen** [comment]
Thanks for the explanation @anandsubbu .  Glad we agree on going after 'static' system information.  That makes sense to me.

What information specifically do you want to capture?  Amount of physical memory? Number of CPU cores?

As is, the script spits out more dynamic information than static, in my opinion.  Maybe the script could filter out some of the other elements that might be more distraction than useful.  For example, on my Mac...
```
$ top -l 1 | head -10 2>/dev/null
Processes: 363 total, 2 running, 361 sleeping, 1735 threads
2017/01/04 12:42:04
Load Avg: 2.19, 1.87, 1.55
CPU usage: 22.54% user, 19.7% sys, 58.38% idle
SharedLibs: 167M resident, 37M data, 45M linkedit.
MemRegions: 129777 total, 5335M resident, 189M private, 1860M shared.
PhysMem: 13G used (2764M wired), 2753M unused.
VM: 1076G vsize, 620M framework vsize, 3733004(0) swapins, 3839813(0) swapouts.
Networks: packets: 7978036/7950M in, 4748808/532M out.
Disks: 7485876/92G read, 2696358/81G written.
```

**Comment by mattf-horton** [comment]
On Mac OS X, the `sysctl` CLI gives this info.
```
sysctl machdep.cpu
```
gives output very similar to /proc/cpuinfo.
There isn't an exact analogue of meminfo, but 
```
sysctl hw.memsize
```
returns the total memory size, which is the main piece of static info desired.  
Other memory-related queries are documented if you do `man sysctl` on a Mac.

**Comment by anandsubbu** [comment]
Thanks a bunch @mattf-horton for the suggestion. I have now changes the commands based on sysctl on Mac and meminfo/cpuinfo on Linux.

**Comment by JonZeolla** [comment]
Jumping in a bit late here.

Please take a look at [my PR](https://github.com/anandsubbu/incubator-metron/pull/1) for a suggested improvement.  Tested it on CentOS 6.8 and Mac 10.12.2.

**Comment by anandsubbu** [comment]
Thank you, @JonZeolla. I have merged your changes into this PR. Please let me know if this is fine, or if you would like to do it separately.

@nickwallen , @mattf-horton are there any other parameters that I should consider capturing?

**Comment by mattf-horton** [comment]
I suppose a `df` or equivalent would be worthwhile. (Only report physical volumes, or perhaps volumes > 5GB capacity.) 

The other major limiter on performance / capacity is intra-cluster Ethernet bandwidth. So that would be nice to report.  But this can be tricky or impossible to read, depending on the NIC hardware and BIOS support, and of course whether you're on a physical machine or a VM. I don't really have any guidance to offer, just a wish, if you have an idea of how to approach it.

**Comment by JonZeolla** [comment]
@anandsubbu you did the real changes here, all I made was a tweak.  We can keep it as-is.

I could go along with adding a `df -h` or similar as well.

**Comment by anandsubbu** [comment]
Thanks @mattf-horton and @JonZeolla, I have now included the output of `df -h` . Here is a sample output on Mac.

```
<snip>
Total System Memory = 16384 MB
Processor Model: Intel(R) Core(TM) i7-4870HQ CPU
Processor Speed: 2.50GHz
Total Physical Processors: 4
Total cores: 4
Disk information:
/dev/disk1     465Gi  338Gi  127Gi    73% 88564601 33279109   73%   / 
```

I did attempt to check for commands to capture the NIC card / interface information, but I have to agree with @mattf-horton that it is non-trivial and can vary widely depending on environments. I considered including the output of `ifconfig` at least, however I was not sure if a user's IP address would fall under sensitive information category--which some might not want having captured/shared in logs (or would they?)

**Comment by mattf-horton** [comment]
Yes, it seems almost impossible on a VM.
I'm not in favor of including ifconfig output.  It doesn't tell you anything about bandwidth, and it produces a lot of info that I think is of negligible value for the purposes of this work item.
I'd say you've got enough here, and should proceed with review process.

**Comment by anandsubbu** [comment]
Thanks @mattf-horton .

@nickwallen @JonZeolla is there any other thing you guys feel we should add to this script? Please let me know.

File: metron-deployment/scripts/platform-info.sh [hunk 1]
```diff
@@ -62,3 +62,39 @@ mvn --version
...
+    sysctl machdep.cpu | grep 'machdep.cpu.brand_string' | cut -d: -f2 | cut -d\@ -f1 | awk '{print "Processor Model:" $0}'
+    sysctl machdep.cpu | grep 'machdep.cpu.brand_string' | cut -d: -f2 | cut -d\@ -f2 | awk '{print "Processor Speed:" $0}'
+    sysctl hw.physicalcpu | cut -d: -f2 | awk '{print "Total Physical Processors:" $0}'
+    sysctl machdep.cpu | grep 'machdep.cpu.core_count' | cut -d: -f2 | cut -d\@ -f2 | awk '{print "Total cores:" $0}'
+    echo "Disk information:"
+    df -h | grep "^/" 
+    ;;
+  bsd*)
+    echo "BSD is not currently supported, unable to detect system resources"
+    ;;
```
Comment (nickwallen) [inline metron-deployment/scripts/platform-info.sh]: @anandsubbu This message might be a little misleading.  Could a user misinterpret this as meaning that Metron does not support the given OS, versus just the platform-info.sh script does not support it?

You could just remove the conditionals for BSD, Windows and Solaris and let it fall through to something like "Unable to detect system resources for "${OSTYPE}".

Comment (JonZeolla) [inline metron-deployment/scripts/platform-info.sh]: Do we have a list of what OSs are supported to run Metron on?  I thought BSD, Windows, and Solaris actually weren't supported on the Metron side.  I know there is [a list for HDP](https://docs.hortonworks.com/HDPDocuments/HDP2/HDP-2.5.3/bk_release-notes/content/ch01s02s01.html), but I don't feel like that applies.

Comment (ottobackwards) [inline metron-deployment/scripts/platform-info.sh]: What the script supports and what Metron supports can be different things, there can be bugs logged for improving the script

**Comment by anandsubbu** [comment]
For all other OS types other than `darwin` and` linux`, I have added a generic log message. Please let me know if there is any other change required at this point.

**Comment by anandsubbu** [comment]
Bump... checking to see if there is any other change required for this pull request.

**Comment by nickwallen** [comment]
+1  Thanks @anandsubbu !  I validated that it works as expected on a Mac.

```
$ incubator-metron-pr-400/metron-deployment/scripts/platform-info.sh
Metron 0.3.0
fatal: Not a git repository (or any of the parent directories): .git
--
ansible 2.0.0.2
  config file =
  configured module search path = Default w/o overrides
--
Vagrant 1.8.1
--
Python 2.7.12
--
Apache Maven 3.3.9 (bb52d8502b132ec0a5a3f4c09453c07478323dc5; 2015-11-10T11:41:47-05:00)
Maven home: /usr/local/Cellar/maven/3.3.9/libexec
Java version: 1.8.0_66, vendor: Oracle Corporation
Java home: /Library/Java/JavaVirtualMachines/jdk1.8.0_66.jdk/Contents/Home/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.12.2", arch: "x86_64", family: "mac"
--
Darwin HW12402.local 16.3.0 Darwin Kernel Version 16.3.0: Thu Nov 17 20:23:58 PST 2016; root:xnu-3789.31.2~1/RELEASE_X86_64 x86_64
--
Total System Memory = 16384 MB
Processor Model: Intel(R) Core(TM) i7-4870HQ CPU
Processor Speed: 2.50GHz
Total Physical Processors: 4
Total cores: 4
Disk information:
/dev/disk1     465Gi  300Gi  165Gi    65% 2836412 4292130867    0%   /
```

**Comment by JonZeolla** [comment]
Per the recent dev thread about author representation, this has code which is a mix of me and anandsubbu.

**Comment by anandsubbu** [comment]
Hi @JonZeolla , sure. I will go ahead and revert the change that you did for the case block, just the if-else block per my earlier change and then you could create a separate PR.  Please let me know if that works.

**Comment by JonZeolla** [comment]
If it's OK with the mentors, I don't mind if you just co-opt that commit.

**Comment by anandsubbu** [comment]
Hi @nickwallen , could you please merge the PR by retaining @JonZeolla as the co-opt for this commit?

**Comment by JonZeolla** [comment]
To clarify, I'm giving permission to merge and squash everything into a commit from @anandsubbu without attributing anything to me.

**Pull request closed without merging**